source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,768 @@
+  photon_handler.insert_global_particles(insertion_points_on_proc,
+                                         global_bounding_boxes,
+                                         photon_properties);
+
+  ConditionalOStream pcout(std::cout,
+                           Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) ==
+                             0);
+  this->print_insertion_info(n_total_photons_to_insert, pcout);
+}
+
+template <int dim>
+void
+RayTracingSolver<dim>::write_output_results(
+  const std::vector<Point<dim>> &points)
+{
+  TimerOutput::Scope t(this->computing_timer, "Output VTU");
+
+  // Create a temporary particle handler on the existing triangulation
+  Particles::ParticleHandler<dim> temp_handler(triangulation, mapping, 0);
+
+  MPI_Comm communicator = triangulation.get_mpi_communicator();
+  // Obtain global bounding boxes
+  const auto my_bounding_box = GridTools::compute_mesh_predicate_bounding_box(
+    triangulation, IteratorFilters::LocallyOwnedCell());
+  const auto global_bounding_boxes =
+    Utilities::MPI::all_gather(communicator, my_bounding_box);
+
+  std::vector<std::vector<double>> photon_properties(points.size());
+
+  // Insert the photons using the points and the assigned properties.
+  temp_handler.insert_global_particles(points,
+                                       global_bounding_boxes,
+                                       photon_properties);
+
+  // Write particles to a VTU file
+  Particles::DataOut<dim> data_out;
+  data_out.build_patches(temp_handler);
+
+  const std::string filename = parameters.simulation_control.output_folder +
+                               "/" + parameters.simulation_control.output_name +
+                               ".vtu";
+
+  data_out.write_vtu_in_parallel(filename, this->mpi_communicator);
+}
+
+// // Flatten local points into a buffer of chars (text format)
+// std::ostringstream oss;
+// if (this_mpi_process == 0)
+//   oss << "x, y, z\n";
+//
+// for (const auto &p : points)
+//   {
+//     oss << p[0] << "," << p[1] << "," << p[2] << "\n";
+//   }
+//
+// std::string local_str  = oss.str();
+// int         local_size = local_str.size();
+//
+// // Gather sizes to compute offsets
+// auto recv_counts = Utilities::MPI::all_gather(mpi_communicator,
+// local_size);
+//
+// std::vector<int> off_set(n_mpi_processes, 0);
+// int              total_size = 0;
+// for (unsigned int i = 0; i < n_mpi_processes; ++i)
+//   {
+//     off_set[i] = total_size;
+//     total_size += recv_counts[i];
+//   }
+//
+// // Build full filename (same for all ranks)
+// std::string full_filename = folder + "/" + file_name + ".xyz";
+//
+// // Open MPI file
+// MPI_File fh;
+// MPI_File_open(mpi_communicator,
+//               full_filename.c_str(),
+//               MPI_MODE_CREATE | MPI_MODE_WRONLY,
+//               MPI_INFO_NULL,
+//               &fh);
+//
+// // Each rank writes at its own offset
+// MPI_File_write_at(fh,
```

Comment:
I would remove this. Or if you feel like. you need it, make it into another function.